// Events / data from the IB API.
package proto.ib;

option optimize_for = LITE_RUNTIME;

import "common.proto";

message MarketData {

  required uint64 timestamp = 1;
  required string symbol = 2;
  required string event = 3;
  required proto.common.Value value = 4;
  required uint64 contract_id = 5;
}

message MarketDepth {

  enum Side {
    ASK = 0;
    BID = 1;
  }

  enum Operation {
    INSERT = 0;
    UPDATE = 1;
    DELETE = 2;
  }

  required uint64 timestamp = 1;
  required string symbol = 2;
  required Side side = 3;
  required double price = 4;
  required uint64 size = 5;
  required Operation operation = 6;
  required uint32 level = 7;
  optional string mm = 8;

  required uint64 contract_id = 9;
}

/** Example twsContract (see R/IBrokers module):
 $ conId          : chr "96099040"
 $ symbol         : chr "GOOG"
 $ sectype        : chr "OPT"
 $ exch           : chr "SMART"
 $ primary        : chr ""
 $ expiry         : chr "20111028"
 $ strike         : chr "590"
 $ currency       : chr "USD"
 $ right          : chr "C"
 $ local          : chr "GOOG  111028C00590000"
 $ multiplier     : chr "100"
 $ combo_legs_desc: chr ""
 $ comboleg       : chr ""
 $ include_expired: chr ""
 $ secIdType      : chr ""
 $ secId          : chr ""
*/
message Contract {

  enum Type {
    INDEX = 0; // IND
    STOCK = 1; // STK
    OPTION = 2; // OPT
  }

  required uint64 id = 1;
  required string symbol = 2;

  optional Type type = 3 [ default = STOCK ];
  optional string exchange = 4 [ default = "SMART" ];
  optional string local_symbol = 5;

  enum Right {
    PUT = 0; // P
    CALL = 1;  // C
  }
  optional Right right = 6;
  optional proto.common.Money strike = 7;
  optional proto.common.Date expiry = 8;
  optional int32 multiplier = 9 [ default = 100 ];

}

message RequestMarketData {
  optional string key = 1 [ default = "IBAPI.RequestMarketData" ];
  required Contract contract = 2;
  optional string tick_types = 3;
  optional bool snapshot = 4 [ default = false ];
}

message CancelMarketData {
  optional string key = 1 [ default = "IBAPI.CancelMarketData" ];
  required Contract contract = 2;
}

message RequestMarketDepth {
  optional string key = 1 [ default = "IBAPI.RequestMarketDepth" ];
  required Contract contract = 2;
  optional uint32 rows = 3 [ default = 10 ];
}

message CancelMarketDepth {
  optional string key = 1 [ default = "IBAPI.CancelMarketDepth" ];
  required Contract contract = 2;
}
